aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -297,19 +297,37 @@ pub struct AggregatorData {
 
 impl AggregatorData {
     /// Returns a mutable reference to an aggregator with `id` and a `limit`.
-    /// If transaction that is currently executing did not initilize it), a new
-    /// aggregator instance is created, with a zero-initialized value and in a
-    /// delta state.
+    /// If transaction that is currently executing did not initilize it, a new aggregator instance is created.
+    /// The state of the new aggregator instance depends on the `aggregator_enabled` flag.
+    /// If the `aggregator_enabled` flag is true, the new aggregator instance
+    /// is initialized with zero and in a delta state.
+    /// If the `aggregator_enabled` flag is false, the new aggregator instance
+    /// is set initialized with its latest value and in a Data state.
     /// Note: when we say "aggregator instance" here we refer to Rust struct and
     /// not to the Move aggregator.
-    pub fn get_aggregator(&mut self, id: AggregatorID, limit: u128) -> &mut Aggregator {
+    pub fn get_aggregator(
+        &mut self,
+        id: AggregatorID,
+        limit: u128,
+        resolver: &dyn TableResolver,
+        aggregator_enabled: bool,
+    ) -> &mut Aggregator {
         self.aggregators.entry(id).or_insert_with(|| Aggregator {
             value: 0,
             state: AggregatorState::PositiveDelta,
             limit,
             history: Some(History::new()),
         });
-        self.aggregators.get_mut(&id).unwrap()
+
+        let aggregator = self.aggregators.get_mut(&id).unwrap();
+        if !aggregator_enabled {
+            let result = aggregator.read_and_materialize(resolver, &id);
```

Comment:
`read_and_materialize` sets the value and the state already, no need  to do `if let`. Maybe let's wrap the return result into PartialVMResult to avoid unwraps then

Reply:
Updated this. Please let me know if it works.